### PR TITLE
Feat/save recalculation object to datastore

### PR DESCRIPTION
**Pull request by deeonwuli**
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/860ru2y5v

### :video_camera: Screenshots/Screen capture
https://github.com/EyeSeeTea/d2-reports/assets/37223065/3da3e640-ac81-4f08-9cf1-00453f43399d

### :fire: Notes to the tester
```
REACT_APP_DHIS2_BASE_URL=https://dev.eyeseetea.com/who-dev-238/
REACT_APP_REPORT_VARIANT=glass-admin
```

### CODE REVIEW

File: src/domain/reports/glass-admin/entities/GLASSDataMaintenanceItem.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    year: string;
+}
+
+export interface ATCItemIdentifier {
+    currentVersion: boolean;
+    version: string;
+    year: string;
+}
+
+export interface AMCRecalculation {
```
Comment (anagperal) [inline src/domain/reports/glass-admin/entities/GLASSDataMaintenanceItem.tsx:46]: This has to be like the following: 
```
export type AMCRecalculation = {
    date: string;
    recalculate: boolean;
    periods: string[];
    orgUnitsIds: Id[];
   loggerProgram: string;
};
```
Please, check that in all the code is `orgUnitsIds` instead of `orgUnitIds` and `date` instead of `currentDate` because this is how it was typed in GLASS app

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+
+    async saveColumns(namespace: string, columns: string[]): Promise<void> {
+        return this.storageClient.saveObject<string[]>(namespace, columns);
+    }
+
+    async uploadATC(namespace: string, file: File, year: string, items: ATCItemIdentifier[]): Promise<void> {
+        const atcItems = await this.getATCItems(namespace);
+        const jsons = await this.extractJsonFromZIP(file);
+
+        if (jsons.length === 4) {
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:107]: Now we change the number of jsons to **6**. `ddd_alterations` and `atc_alterations` jsons have been added together with `ddd`, `atc`, `ddd_combinations` and `conversion`

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+            const updatedRecalculationLogic = {
+                ...amcRecalculationLogic,
+                recalculate: false,
+            };
+
+            await this.globalStorageClient.saveObject<AMCRecalculation>(namespace, updatedRecalculationLogic);
+        }
+    }
+
+    async saveRecalculationLogic(namespace: string, atcNamespace: string): Promise<void> {
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:142]: This function does not fulfill the solid principle of single responsibility: the name of the function makes me think that only amc-recalculations is saved in the data store, but you are also updating ATCs. It's better to split here in the repository in two different functions and both functions are called from the use case that is in charge of coordinating them. Also rename the use case according to what it has to do

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+        const amcRecalculationLogic = await this.getRecalculationLogic(namespace);
+        const atcs = await this.getATCItems("ATCs");
+
+        const updatedATCs: ATCItem[] = atcs.map(atc => {
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:146]: Could be simplify like:   
```
        const updatedATCs: ATCItem[] = atcs.map(atc => {
            return {
                ...atc,
                previousVersion: atc.currentVersion,
            };
        });
```

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+                },
+            })
+            .getData();
+
+        return programs[0]?.name ?? "";
+    }
+
+    private getATCPeriods = (): string[] => {
+        const currentYear = new Date().getFullYear();
+        const startYear = 2016;
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:198]: Maybe it's better to create a constant in the upper part of the code named like `const START_YEAR_PERIODS = 2016;` to make it easier to locate and better understand that it is the first year used in the project to start the periods.

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+                    },
+                    paging: false,
+                },
+            })
+            .getData();
+
+        return _.concat(organisationUnits, kosovoOu).map(orgUnit => orgUnit.id);
+    }
+
+    private uploadNewATC(year: string, atcItems: ATCItem[]): ATCItem[] {
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:299]: take into account here what I have said about `currentVersion` and `previousVersion` above, because `currentVersion` may be changed to true here

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+            version: "1",
+            uploadedDate: new Date().toISOString(),
+            currentVersion: _.parseInt(year) > currentVersionYear,
+            previousVersion: false,
+        };
+
+        return [...atcItems, newItem];
+    }
+
+    private patchATCVersion(atcItems: ATCItem[], selectedItems: ATCItemIdentifier[]) {
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:314]: Also here take into account what I have said about `currentVersion` and `previousVersion` above, because `currentVersion` may be changed to true here

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+import { Config } from "../../../domain/common/entities/Config";
+import { Id } from "../../../domain/common/entities/Base";
+import {
+    PaginatedObjects,
+    Paging,
+    Sorting,
+    getPaginatedObjects,
+} from "../../../domain/common/entities/PaginatedObjects";
+
+interface ATCJson {
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:34]: Try to avoid any in types. For this type you can take a look into glass repository at the end of [this file](https://github.com/EyeSeeTea/glass-dev/blob/38ef6e5c1fa16e614eb9855a7d24c067966fa877/src/data/repositories/GlassATCDefaultRepository.ts)

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+        const atcItems = await this.getATCItems(namespace);
+        const jsons = await this.extractJsonFromZIP(file);
+
+        if (jsons.length === 4) {
+            if (items) {
+                const updatedVersion = this.updateVersion(items);
+                const updatedATCItems = this.patchATCVersion(atcItems, items);
+
+                await this.globalStorageClient.saveObject<ATCJson>(`ATC-${updatedVersion}`, jsons);
+                await this.globalStorageClient.saveObject<ATCJson>(Namespaces.ATCS, updatedATCItems);
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:113]: Here you are saving `updatedATCItems` whose type would be ATCItem[]

File: src/webapp/reports/glass-admin/glass-admin-list/atc-classification/useATCUpload.tsx [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    const [isPatchModalOpen, { enable: openPatchModal, disable: closePatchModal }] = useBooleanState(false);
+    const [isUploadATCModalOpen, { enable: openUploadATCModal, disable: closeUploadATCModal }] = useBooleanState(false);
+    const [isRecalculateLogicModalOpen, { enable: openRecalculateLogicModal, disable: closeRecalculateLogicModal }] =
+        useBooleanState(false);
+
+    const [recalculationLogic, setRecalculationLogic] = useState<AMCRecalculation>();
+    const [loggerProgram, setLoggerProgram] = useState<string>("");
+
+    useEffect(() => {
```
Comment (anagperal) [inline src/webapp/reports/glass-admin/glass-admin-list/atc-classification/useATCUpload.tsx:18]: Here instead of first getting data in `amc-recalculation` to just use the `loggerProgram` value to then take the name of the program (which is what is shown), it's better to get in the use case `GetATCLoggerProgramUseCase` the `loggerProgram` from data in `amc-recalculation` and with that get the name of the program which is returned

File: src/webapp/reports/glass-admin/glass-admin-list/atc-classification/useATCActions.tsx [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            .then(() => setIsRecalculated(false));
+        reload();
+        snackbar.success("Recalculation has been cancelled successfully");
+    }, [compositionRoot.glassAdmin, reload, snackbar]);
+
+    const saveRecalculationLogic = useCallback(async () => {
+        try {
+            setIsRecalculating(true);
+            await compositionRoot.glassAdmin.saveRecalculationLogic(Namespaces.AMC_RECALCULATION, Namespaces.ATCS);
+            snackbar.success(`Please go to the program ${loggerProgram} to see the logs of this recalculation`);
```
Comment (anagperal) [inline src/webapp/reports/glass-admin/glass-admin-list/atc-classification/useATCActions.tsx:67]: If here it's the only place where you need `loggerProgram` to show the name in the message, it's better to move  the request from useATCUpload hook and used it like `const loggerProgram = await compositionRoot.glassAdmin.getATCLoggerProgram();`

File: src/webapp/reports/glass-admin/glass-admin-list/atc-classification/ATCClassificationList.tsx [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+                onSave={saveRecalculationLogic}
+                onCancel={closeRecalculateLogicModal}
+                saveText={i18n.t(isRecalculating ? "Recalculating..." : "Start recalculation")}
+                cancelText={i18n.t("Cancel")}
+                disableSave={isRecalculating}
+                maxWidth="md"
+                fullWidth
+            >
+                {previousVersionExists && (
+                    <p>Last recalculation was done with {getVersion(tableProps.rows, "previousVersion")}</p>
```
Comment (anagperal) [inline src/webapp/reports/glass-admin/glass-admin-list/atc-classification/ATCClassificationList.tsx:169]: Use translations in this lines. See example:
i18n.t("Last recalculation was done with  {{previousVersion}}", { previousVersion: getVersion(tableProps.rows, "previousVersion") })

File: src/webapp/reports/glass-admin/glass-admin-list/atc-classification/useATCActions.tsx (hunk 11 above)
Comment (anagperal) [inline src/webapp/reports/glass-admin/glass-admin-list/atc-classification/useATCActions.tsx:67]: `i18n.t("Please go to the program {{loggerProgram}} to see the logs of this recalculation", { loggerProgram) })`

File: src/webapp/reports/glass-admin/glass-admin-list/atc-classification/ATCClassificationList.tsx [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+    );
+});
+
+const StyledButtonContainer = styled.div`
+    display: flex;
+    justify-content: end;
+    gap: 1rem;
+`;
+
+const CancelButton = styled(Button)`
```
Comment (anagperal) [inline src/webapp/reports/glass-admin/glass-admin-list/atc-classification/ATCClassificationList.tsx:184]:
```
const CancelButton = styled(Button)`
    background-color: #f44336;
    color: white;
    &:hover {
        background-color: #d32f2f;
    }
`;
```

**Review by anagperal — CHANGES_REQUESTED** [review]
thanks @deeonwuli ! I have added comments to the code

File: src/data/reports/data-quality/DataQualityDefaultRepository.ts [hunk 14]
```diff
@@ -167,29 +167,25 @@ export class DataQualityDefaultRepository implements DataQualityRepository {
...
-                        numeratorResult,
-                        denominatorResult,
-                        metadataType: "Indicator",
-                    };
-                } catch (error) {
-                    console.debug(error);
-                }
-            })
-        ) as IndicatorItem[];
+        return (await promiseMap(metadataItems, async item => {
```
Comment (anagperal) [inline src/data/reports/data-quality/DataQualityDefaultRepository.ts]: This is not related to GLASS, right?

Comment (anagperal) [reply src/data/reports/data-quality/DataQualityDefaultRepository.ts]: Oh, ignore this comment, I just saw that the following commit is a revert of this

Comment (deeonwuli) [reply src/data/reports/data-quality/DataQualityDefaultRepository.ts]: no. i will revert the changes here

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 15]
```diff
@@ -0,0 +1,499 @@
...
+        const columns = await this.storageClient.getObject<string[]>(namespace);
+
+        return columns ?? [];
+    }
+
+    async saveColumns(namespace: string, columns: string[]): Promise<void> {
+        return this.storageClient.saveObject<string[]>(namespace, columns);
+    }
+
+    async uploadATC(namespace: string, file: File, year: string, items: ATCItemIdentifier[]): Promise<void> {
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:106]: better `items` named as `selectedItems` and add `?` like in `GLASSDataMaintenanceRepository` (also change this on GLASSDataMaintenanceRepository )

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts [hunk 16]
```diff
@@ -0,0 +1,499 @@
...
+            .value();
+    }
+
+    private patchATCVersion(atcItems: ATCItem[], selectedItems: ATCItemIdentifier[]): ATCItem[] {
+        return atcItems.reduce((acc: ATCItem[], atcItem: ATCItem) => {
+            const matchingItem = selectedItems.find(
+                item => item.year === atcItem.year && _.parseInt(item.version) === _.parseInt(atcItem.version)
+            );
+
+            if (matchingItem) {
```
Comment (anagperal) [inline src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:324]: We have forgotten more logic here:

1. If `matchingItem` is the `currentVersion`: we should change the new patched version to `currentVersion: true`, change `matchingItem` to `previousVersion: true` and look for the one with `previousVersion: true` to change it to `false` since it will no longer be `true`
2. If `matchingItem` is not the `currentVersion` and is not the `previousVersion`: what should we do? Will this new patched version be the one to be used in the calculations? the new patched version should now be the `currentVersion` (and therefore the previous `currentVersion` will be the `previousVersion` and the `previousVersion` will no longer be this)??
3. If `matchingItem` is the `previousVersion`: what should we do? Will this new patched version be the one to be used in the calculations? the new patched version should now be the `currentVersion` (and therefore the previous `currentVersion` will be the `previousVersion` and the `previousVersion` will no longer be this)??

**Review by anagperal — CHANGES_REQUESTED** [review]
thanks @deeonwuli ! nice solutions to PR comments! :tada: I've added two more related to patch ATC version

File: src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts (hunk 16 above)
Comment (anagperal) [reply src/data/reports/glass-admin/GLASSDataMaintenanceDefaultRepository.ts:324]: What makes sense to me it's the next logic:
1. If `matchingItem` is the `currentVersion`: we should change the new patched version to `currentVersion: true`, change `matchingItem` to `previousVersion: true` and look for the one with `previousVersion: true` to change it to `false` since it will no longer be `true`
2. Any other situation, simply add the new patched version with `previousVersion: false` and `currentVersion: false`

**Comment by deeonwuli** [comment]
> thanks @deeonwuli ! Code looks good. I have tested the functionality a bit and I have found two possible problems:
> 
> 1. Both when I upload a new zip and when I make a patch, the modal allows me to click more than once on the continue button because it is making the request from behind and the modal stays open. You could do like in the recalculate modal that you set the button to disable once you have clicked continue. **The problem is that in one of the tests when I clicked it twice it caused it to create two patched versions.**
> 2. The cancel recalculation button disappears when I reload the page even though in the data store it is still in amc-recalculate the `recalculate: true`. This is because you are using a state with init state `false`, but I think that this init state should depend on the value from the amc-recalculate to allow the user change `recalculate` from `true` to `false` as long as you have `true`

great catch @anagperal! thanks for pointing these out 😁

**Review by anagperal — APPROVED** [review]
thanks @deeonwuli !!

**Review by ifoche — APPROVED** [review]
thanks @deeonwuli
